File: apps/mobile/lib/crypto/message_crypto.dart
// ============================================================
//  Message Crypto
//  AES-256-GCM encryption/decryption for message payloads.
//  Format: base64(nonce || ciphertext || tag)
// ============================================================

import 'dart:convert';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';

class MessageCrypto {
  static final _aesGcm = AesGcm.with256bits();

  /// Encrypt plaintext with a shared secret (Base64 key).
  /// Returns a single Base64 string: nonce || ciphertext || tag.
  static Future<String> encrypt(String plaintext, String sharedSecretB64) async {
    final keyBytes = base64Decode(sharedSecretB64);
    final secretKey = SecretKey(keyBytes);
    final plaintextBytes = utf8.encode(plaintext);

    final secretBox = await _aesGcm.encrypt(
      plaintextBytes,
      secretKey: secretKey,
    );

    // Combine: nonce (12 bytes) + ciphertext + mac (16 bytes)
    final combined = Uint8List.fromList([
      ...secretBox.nonce,
      ...secretBox.cipherText,
      ...secretBox.mac.bytes,
    ]);

    return base64Encode(combined);
  }

  /// Decrypt a Base64 ciphertext string with a shared secret (Base64 key).
  /// Input format: base64(nonce || ciphertext || tag).
  static Future<String> decrypt(String ciphertextB64, String sharedSecretB64) async {
    final keyBytes = base64Decode(sharedSecretB64);
    final secretKey = SecretKey(keyBytes);
    final combined = base64Decode(ciphertextB64);

    // Extract nonce (12 bytes), ciphertext, and mac (16 bytes)
    final nonce = combined.sublist(0, 12);
    final mac = Mac(combined.sublist(combined.length - 16));
    final cipherText = combined.sublist(12, combined.length - 16);

    final secretBox = SecretBox(
      cipherText,
      nonce: nonce,
      mac: mac,
    );

    final decrypted = await _aesGcm.decrypt(
      secretBox,
      secretKey: secretKey,
    );

    return utf8.decode(decrypted);
  }

  /// Generate a random AES-256 key, returned as Base64.
  static Future<String> generateRandomKey() async {
    final secretKey = await _aesGcm.newSecretKey();
    final keyBytes = await secretKey.extractBytes();
    return base64Encode(Uint8List.fromList(keyBytes));
  }
}

File: apps/web/src/lib/crypto/message-crypto.ts
// ============================================================
//  Web Crypto — Message Encryption/Decryption
//  AES-256-GCM using Web Crypto API.
//  Format: base64(nonce || ciphertext || tag)
// ============================================================

function toBase64(buffer: ArrayBuffer): string {
  return btoa(String.fromCharCode(...new Uint8Array(buffer)));
}

function fromBase64(b64: string): ArrayBuffer {
  const binary = atob(b64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes.buffer as ArrayBuffer;
}

/**
 * Encrypt plaintext with AES-256-GCM.
 */
export async function encrypt(
  plaintext: string,
  sharedSecretB64: string
): Promise<string> {
  const keyBuffer = fromBase64(sharedSecretB64);
  const key = await crypto.subtle.importKey(
    "raw",
    keyBuffer,
    { name: "AES-GCM" },
    false,
    ["encrypt"]
  );

  const nonce = crypto.getRandomValues(new Uint8Array(12));
  const plaintextBytes = new TextEncoder().encode(plaintext);

  const encrypted = await crypto.subtle.encrypt(
    { name: "AES-GCM", iv: nonce },
    key,
    plaintextBytes
  );

  const encryptedArray = new Uint8Array(encrypted);
  const combined = new Uint8Array(nonce.length + encryptedArray.length);
  combined.set(nonce, 0);
  combined.set(encryptedArray, nonce.length);

  return toBase64(combined.buffer as ArrayBuffer);
}

/**
 * Decrypt AES-256-GCM ciphertext.
 */
export async function decrypt(
  ciphertextB64: string,
  sharedSecretB64: string
): Promise<string> {
  const keyBuffer = fromBase64(sharedSecretB64);
  const key = await crypto.subtle.importKey(
    "raw",
    keyBuffer,
    { name: "AES-GCM" },
    false,
    ["decrypt"]
  );

  const combined = new Uint8Array(fromBase64(ciphertextB64));
  const nonce = combined.slice(0, 12);
  const ciphertext = combined.slice(12);

  const decrypted = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv: nonce },
    key,
    ciphertext.buffer as ArrayBuffer
  );

  return new TextDecoder().decode(decrypted);
}

/**
 * Generate a random AES-256 key.
 */
export async function generateRandomKey(): Promise<string> {
  const key = await crypto.subtle.generateKey(
    { name: "AES-GCM", length: 256 },
    true,
    ["encrypt", "decrypt"]
  );
  const raw = await crypto.subtle.exportKey("raw", key);
  return toBase64(raw);
}
